### ISSUE TITLE
liblink, cmd/ld: don't encode the instruction being relocated in Reloc.add

### ISSUE THREAD

**Issue by minux** [golang/go#10050]
When I first implemented cgo for ARM, I made the wrong precedence to encode
the instruction itself in the relocation's addend field. It makes everything harder:

liblink must stuff the real addend into the instruction, put it in reloc.add, and then
in the linker, it must get the real addend out from the instruction, calculate the
real value, and stuff the value back into the instruction.

Additionally, it makes -S output unreadable.
```
rel 20+4 t=6 runtime.morestack_noctxt+9bfffffe
rel 40+4 t=6 runtime.duffzero+eb00005e
```
(Note: the 2nd line is calling into the middle of duffzero, but you can't tell
anything from the addend without some shifts and masks to figure out the
offset)
Compare this to the gg -S output:
```
rel 16+4 t=4 runtime.morestack_noctxt+0
rel 37+4 t=4 runtime.duffzero+228
```

I suspect this is because in cmd/ld/lib.c, we always pass a zero val (the variable o)
to archreloc, instead, it really should pass the original value read from the symbol.

Should we correct this? so that archreloc is passed the real instrutions/data being
relocated in *val argument, and we place the real addend in Reloc.add field?

**Comment by mwhudson** [comment]
I don't know how seriously my input should be taken here, but this whole game confused the heck out of me when I did arm64 external linking.  Passing the original value to archreloc would make a lot more sense.

**Comment by minux** [comment]
Yeah, I'm trying to add R_ADDRARM64, and I really can't bear the current
scheme anymore.
It's just too much unnecessary work (and ADRP encodes the immediate value
in two parts
make everything even more annoying.)

**Comment by josharian** [comment]
FWIW, this threw me for a while today.

**Comment by rsc** [comment]
Too late for Go 1.5.
